### DISCUSSION TITLE
Barriers / Gates in BRouter / Graphhopper

### DISCUSSION THREAD

**Discussion by MotoUKRider** [devemux86/cruiser#56] [Q&A] [answered]
This isn't really a Cruiser issue, but details a few tweaks I've made to the car-fast.brf BRouter profile in order to make it work a little better for me.

I'm in quite a rural part of the UK, there are quite a few cattle-grids and level-crossings.  Previously, I had wondered why I'd be taken on a 10-mile detour between 2 Via-Points that were within visual range on the same road.  Turns out that BRouter (and GraphHopper) won't take you over a cattle-grid - and, if somebody has been editing OpenStreetMap in their own way: A standard level-crossing with an automatic barrier would also be avoided if there was a "lift_gate" at the location.  Similar crossings locally confirm that the element placements are *not* consistent (the joys of community edits!).  I am able to route over a crossing a few miles away. Meh.

The BRouter mod is as follows:
Find (approx line 250): `not barrier=gate|bollard|lift_gate|cycle_barrier`
Change to: `not barrier=bollard|cycle_barrier`
*(I made a copy of car-fast.brf and renamed it so that I still had the original)*

I don't know much about the map tags, so am not sure what else a "gate" would mean in other places, perhaps an actual physical barrier/gate for cycle/pedestrian access only?  Who knows - but it certainly applies to at least 5 cattle-grids that I've got on some of my routes.

**Comment by MotoUKRider** [comment]
Here's an example that I found within Cruiser Desktop.  The routing is fine via online BRouter and OpenRouteService, but fails miserably using offline GraphHopper - it tries to take me down a footpath which then becomes a road (and back the same way) despite "Car" being selected.

BRouter/OpenRouteService Route:
![route-issue-gates-1](https://user-images.githubusercontent.com/125696109/221237261-82a07c4b-ad7e-43c4-8dc3-e420a1d0b980.jpg)

GraphHopper Route:
![route-issue-gates-2](https://user-images.githubusercontent.com/125696109/221237270-0458246f-0982-467b-8c7a-4c7ec281dd4e.jpg)

The following objects appear to be the issue (they are cattle-grids):
https://www.openstreetmap.org/node/2923770193
https://www.openstreetmap.org/node/3284680573

It's clearly a routing issue but hopefully provides some insight into why things happen and how to avoid them - or, in my case, be able to ride over them :)

**Comment by devemux86** [comment]
@MotoUKRider thanks for these examples and the BRouter profile instructions! :+1:

**Comment by 0709wiwiwi** [comment]
@MotoUKRider.

Your contributions regarding BRouter profiles optimised for the motorcyclist target group are valuable. Also know that BRouter is mainly aimed at cyclists and hikers. Thus, optimal motorbike profiles have been neglected anyway. However, your contributions in the Cruiser forum will probably be lost too soon anyway.

Since you are pretty good with profile building anyway, perhaps you could share your ideas and experience regarding a better motorbike router profile in Brouter github. 
Possibly your optimal profile could one day be available by default together with a Brouter app download. 

Good luck.


Translated with www.DeepL.com/Translator (free version)

**Reply by MotoUKRider** [reply]
Thanks for the nice words - I'm not sure I have anything to contribute in that regard.  My riding via Cruiser is via a planned route that I want to follow (shaping-points, places to stop etc) rather than "A to B, no motorways, very curvy" - there's at least 2 better apps if that's your thing.

The routing issue with GraphHopper is more of a concern: If I select "Bike", the route is acceptable. It's almost as if Cruiser is telling GraphHopper the incorrect vehicle.

**Reply by devemux86** [reply]
> The routing issue with GraphHopper is more of a concern: If I select "Bike", the route is acceptable. It's almost as if Cruiser is telling GraphHopper the incorrect vehicle.

GraphHopper does not have a separate motorcycle internal profile.
It uses its car internal profile.

**Reply by MotoUKRider** [reply]
"Car" was used for the examples above. If I select "Bike" in GraphHopper, it plotted the correct/short route (like BRouter).  It would seem that the "barrier/gate" tag is causing an issue, I'm guessing a width restriction or physical barrier that the routing algorithm denies a motor-vehicle access to?

In short, I'm thinking that some overrides may be required to allow for motorcycle routing where we *know* it's safe to ride. I wonder if other examples exist within Europe that can be tested?

![cattle-grid](https://user-images.githubusercontent.com/125696109/221367011-013de979-ea65-4b65-9625-6559b08696b0.jpg)

**Comment by MotoUKRider** [comment]
More investigations - I'm afraid that Graphhopper, in the UK at least, is a bit rubbish!  The presence of these cattle-grids and those types of "barrier" level-crossings make it impossible to route a car or motorcycle profile through them. There must be examples like this within Europe that pose similar hurdles?

Even BRouter in its default "carfast" configuration has this problem, hence my custom profile (details in 1st post).  The use of `gate` and `lift_gate` seem to be incorrectly applied to various road "furniture" on OSM - certainly not with any degree of consistency. Does this mean that the map tags are wrong?  Does this mean that there is a tag that is yet to be defined to accurately detail the road item?  Does this mean that community edits need triple-checking by a local user and/or StreetView?  Only Smarties have the answer!

It's my opinion that Graphhopper is routing incorrectly: Preventing vehicles from passing through "barriers". I've checked a few local ones and the "access" tag isn't present.  I took a read of [OSM Access Tags](https://wiki.openstreetmap.org/wiki/OSM_tags_for_routing/Access_restrictions) and ended up in a corn maze!  There appear to be all kinds of variables for access: Time, vehicle type etc and whatever implied rights you have based on the Country you're in. 

Anywhere there is farm land and/or level-crossings on your route may cause Graphhopper routing to do all kinds of interesting things. This isn't Cruiser's fault, of course, but it does pose a problem in the above scenario - and worth considering especially as Graphhopper is "built-in" and possibly the go-to routing method for offline use.  If it's possible for Cruiser to tell the routing engine what kind of road/ferry/gate/toll are avoidances/preferences then it may be possible to create better "Car" and "Motorcycle" profiles (plus custom ones) where you can avoid whatever attributes/roads/tags are on offer.

Or is it just me?!

**Reply by 0709wiwiwi** [reply]
No.
https://forum.kurviger.de/search?expanded=true&q=Barriers

**Reply by MotoUKRider** [reply]
Oh wow!  Think I may need a "Turbo Boost" button on the bike :)

**Reply by MotoUKRider** [reply]
I wouldn't mind so much if this type of stuff was easily configurable (for the basic user) - It shouldn't fall to editing files within Notepad, just checkboxes in a Routing Profile within an app (most nav apps have this). That said, it's still "not quite right" that a car routing profile will not take me over a cattle-grid (not-to-mention the many variables used in tagging map features from the community).

**Reply by devemux86** [reply]
@MotoUKRider 

I agree, that's why I want to explore how we can better handle BRouter profiles / options inside Cruiser.

**Reply by MotoUKRider** [reply]
To be fair, BRouter isn't the main culprit - although it's certainly more work for the average user to maintain download zones and decide from a list of generic profiles (never mind editing them).  Graphhopper is the main issue in that there doesn't seem to be a way tell it how to handle the routing issues outlined above. If the API allows Cruiser to pass detailed criteria, then it suggest that this can be resolved without nagging the Graphhopper forum :)

The use of certain tags (barrier, gate, troll) is either incorrect (due to community edits) or the routing algorithms are using them incorrectly.

**Reply by devemux86** [reply]
OsmAnd handles this nicely and it was my [suggestion](https://forum.kurviger.de/t/consider-temporary-limitations/6442/2) in the past, but it wasn't accepted.

A global on/off switch for temporary limitations so routes can be calculated
and possibly include info for limits in route response so apps can display it.

<img src="https://user-images.githubusercontent.com/3484020/228477314-f223338e-e6ec-4c53-bad8-54b6afc21e18.png" width="33%"/>

**Comment by afischerdev** [comment]
@0709wiwiwi 
In your sample the definition is very clear:
```
access=no
barrier=gate
```

@MotoUKRider 
cattle_grid and BRouter: I searched a little bit and had this [result](https://brouter.de/brouter-web/#map=16/50.8507/-1.7427/standard&lonlats=-1.733984,50.852667;-1.751674,50.845406&profile=car-eco)
Means to me, for cars cattle_grid is passable - without changes in profile. 
May be you have a sample for your detour.

**Reply by MotoUKRider** [reply]
Sorry, let me be clearer:

Graphhopper is the main issue re: cattle-grids. Can't route, won't route :(

BRouter doesn't like certain level-crossings, eg: [Diverted from Level-Crossing](https://brouter.de/brouter-web/#map=13/53.5263/-0.4908/standard&lonlats=-0.520949,53.544464;-0.528213,53.508474&profile=car-fast&alternativeidx=1) which is quite odd. I tested this in the web interface and Cruiser app - and only my edited BRouter profile will route me *direct* over *that* crossing.  The [tag of crossing:barrier = Half](https://www.openstreetmap.org/node/691317383) could be the reason why.

Here's [another example](https://brouter.de/brouter-web/#map=13/53.5816/-0.4456/standard&lonlats=-0.471511,53.561036;-0.445633,53.602266&profile=car-fast&alternativeidx=1), but it *does* take me over it.  There is [no barrier attribute](https://www.openstreetmap.org/node/691317696#map=19/53.58078/-0.46145) on this crossing.

What I think is that there is no (routing) difference between a level-crossing barrier and a car-park barrier (to private/public parking).

**Comment by devemux86** [comment]
> The following objects appear to be the issue (they are cattle-grids):
https://www.openstreetmap.org/node/2923770193
https://www.openstreetmap.org/node/3284680573

@MotoUKRider 

I try your 2 examples on BRouter website with car profiles and they work.
Has something changed?

**Reply by MotoUKRider** [reply]
Ah, the gate/cattle-grid appears to be OK thanks - It's the level-crossing `barrier` tag using Cruiser app and offline (lastest public release) BRouter's car-fast profile.  BRouter website also produces same issue.

See https://github.com/devemux86/cruiser/discussions/56#discussioncomment-5463651 where I mention 2 examples of similar crossings where one routes OK and the other does not.  I have to use my editing BRouter profile within Cruiser app to pass over the crossing on the B1206.

`gate` seems to be the issue in Graphhopper - with my Scottish example (and picture of the road) in the other thread taking me on a long detour down a single-track gravel road (only each end visible on GoogleMaps), so it's passable, but not ideal.

**Reply by devemux86** [reply]
@MotoUKRider thanks for the explanation.

I asked @afischerdev if / how we can use such _internal_ profile parameters (like 'barrier') in external BRouter calls.

It would be nice if we could have a switch for "temporary limitations", as it works elsewhere:
- https://github.com/devemux86/cruiser/discussions/56#discussioncomment-5463073

**Reply by devemux86** [reply]
> I asked @afischerdev if / how we can use such internal profile parameters (like 'barrier') in external BRouter calls.

It does not seem possible, we need a modified profile for such parameters.

**Reply by MotoUKRider** [reply]
OK - it was worth an ask, at least!

This does re-ignite the issue regarding consistency of tagging road features/furniture - and (new/non-expert) users being at the mercy of a custom profile (possibly 5-pages deep on a forum) to provide a workaround and perhaps, unfairly, blaming the navigation app instead of the routing engine :(

The main thing is that the issue can be worked-around, and perhaps we can add some "primer" type help/walkthrough to the Wiki page or I can re-publish my BRouter profile again?

I'm sure you'll agree that requiring Mr/Mrs Average User to edit a text file isn't ideal.

**Comment by devemux86** [comment]
@MotoUKRider about your report in GraphHopper:
- https://github.com/graphhopper/graphhopper/issues/2816

Testing your example:
- GraphHopper **online** should work properly inside Cruiser?
- GraphHopper **offline** may be more restricted.
I will see if we can improve it in the future.

**Reply by MotoUKRider** [reply]
Yes, the offline graphs in Desktop+App take me on a wild-west tour!

**Reply by devemux86** [reply]
On desktop better use the routing websites, export to gpx format and import to Cruiser.
Websites usually offer more routing options and cross-border routing.

**Reply by MotoUKRider** [reply]
It would be nice *not* to have to rely on the interwebs!
Online Graphhopper is OK, so the offline file (or the API and parameters sent to/from it are the issue).

Could you investigate the Cruiser<>Graphhopper exchange of parameters because I think that's the real issue here.  If there are "avoidances" that can be specified, it would be useful to at least have a config option.  The infamous cattle-grid is not any kind of physical barrier or preventative routing object - its only caution is that it's slippery when wet :)

**Comment by devemux86** [comment]
@afischerdev suggested to include the BRouter .brf profiles inside Cruiser.
Add a global variable in these profiles to switch on/off such internal options.

Then I can add a switch in Cruiser settings to turn on/off barriers, gates, ...
Similar to the current Cruiser avoidance settings for BRouter.

Has anyone experience with the use of such global variables in BRouter?
Do you have an example profile to test?

**Comment by devemux86** [comment]
> Could you investigate the Cruiser<>Graphhopper exchange of parameters

Ignore barriers in GraphHopper (offline) is possible, stay tuned for the next version!

Ignore barriers in BRouter (offline) will be handled separately:
- https://github.com/devemux86/cruiser/discussions/56#discussioncomment-5839071

**Comment by devemux86** [comment]
- **BRouter (offline)**: see https://github.com/devemux86/cruiser/discussions/175

- **GraphHopper (offline)** now works similar to the GraphHopper website:
_Update your GraphHopper downloads!_

**Reply by MotoUKRider** [reply]
Thanks for this - I can now use GraphHopper routing for my adventurous Scotland trips!  Only the cows+sheep can't cross :)

**Reply by MotoUKRider** [reply]
@devemux86 Sorry to drag this up, but in my (extensive) use of Desktop for the theme work, I (re)downloaded a couple of GraphHopper areas and the UK one seems to have re-introduced [that "cattle_grid" issue](https://github.com/graphhopper/graphhopper/issues/2816) again. The [infamous Scottish cattle-grid](https://www.openstreetmap.org/node/2923770193#map=19/55.36555/-3.85018) is no longer allowing me to pass.

Using the online GraphHopper Route Planner, there is no issue with the cattle-grid.

There remains an issue with certain types of level-crossing "barrier types" (deja vu much?!).  A defined `crossing_barrier` type of `half` trips it up, but when a crossing has no barrier definition, it's allows access/passage.

I also won't tell anyone that it took me several Desktop archive versions, folder open attempts and coffee before I realised that you had to *unzip* the routing files before they could be used. It's amazing how much you forget when you rely on another method (online BRouter, in this case).

**Reply by devemux86** [reply]
@MotoUKRider thanks the report.

The graphs seem to not be correct and have only the car profile.
For example the scotland graph from [here](https://www.dropbox.com/scl/fo/0pi1svwz14rl49l2gldqr/ANVb1vCX3ezM-cuXr532fFY?rlkey=3lt3t50tinxdsdkfz0py1t4ec&st=e2snb8iu&dl=0) should work correctly.

We will need to rebuild the graphs in the server.

**Reply by MotoUKRider** [reply]
@devemux86 Thanks for the file - yes, that routing ZIP fixes the cattle-grid issue.

**Reply by devemux86** [reply, marked as answer]
@MotoUKRider 

GraphHopper (offline) routing data have all been recreated.

**Reply by MotoUKRider** [reply]
@devemux86 Thanks - that cures both my known trouble-spots!

**Comment by MotoUKRider** [comment]
Wow, it was over 2 years ago that we had "fun" with cattle-grids in the routing engines...

This week, I was finally able to visit [the offending road feature further up this thread](https://github.com/devemux86/cruiser/discussions/56#discussioncomment-5109074) during my trip into Scotland.

I present the Mennock Pass cattle-grid *before* I rode across it singing: _"nerr-nerr-na-nerr-nerrrr"_
![vstromxt-cruiser-cattle-grid-2](https://github.com/user-attachments/assets/e97595c3-f214-4d23-93f6-126c8dea54cd)

From here the road leads up to Wanlockhead, the highest village in Scotland (466m) and I can assure you the views are not 'arf bad!

**Reply by devemux86** [reply]
@MotoUKRider Thanks for the picture!

We did so many routing tests with this particular area. :slightly_smiling_face:
